gapis/resolve/command_tree.go

```diff
@@ -72,30 +72,28 @@ func (t *commandTree) index(indices []uint64) (api.SpanItem, api.SubCmdIdx) {
 	return group, subCmdRootID
 }
 
-func (t *commandTree) indices(idx []uint64, preferGroup bool) []uint64 {
+func (t *commandTree) indices(cmdIdx []uint64, preferGroup bool) []uint64 {
 	out := []uint64{}
 	group := t.root
 
-	for _, id := range idx {
-		brk := false
-		for {
-			if brk {
-				break
-			}
-			i := group.IndexOf(api.CmdID(id))
-			out = append(out, i)
-			switch item := group.Index(i).(type) {
-			case api.CmdIDGroup:
-				group = item
-				if preferGroup {
-					brk = true
-				}
-			case api.SubCmdRoot:
-				group = item.SubGroup
-				brk = true
-			default:
-				return out
+	for i := 0; i < len(cmdIdx); i++ {
+		cmdId := api.CmdID(cmdIdx[i])
+		idx := group.IndexOf(cmdId)
+		out = append(out, idx)
+
+		switch item := group.Index(idx).(type) {
+		case api.CmdIDGroup:
+			group = item
+			if group.Bounds().Last() > cmdId {
+				// We are looking for something inside this group, not the group itself, so go around again.
+				i--
```

Comment:
OK so this is confusing. So the cmdIds within the tree are ordered? How is decreasing ```i``` going to get you inside the group?

Reply:
Decreasing `i` makes the loop go around again in the same "place", but we've updated `group` to point to the next group, so the it will be processed next. This loop is a bit confusing, because it's mixing an iterative pattern with a recursive pattern. It was the same before with two nested loops. This part didn't change and since we'll refactor it later, I didn't want to change this too much.